components/org.wso2.carbon.identity.authz.service/src/main/java/org/wso2/carbon/identity/authz/service/handler/AuthorizationHandler.java

```diff
@@ -104,6 +111,23 @@ public AuthorizationResult handleAuthorization(AuthorizationContext authorizatio
             } else {
                 AuthenticatedUser authenticatedUser = new AuthenticatedUser(user);
                 String userId = PrivilegedCarbonContext.getThreadLocalCarbonContext().getUserId();
+                // Check whether the user is accessing a resource where the user has the access.
+                String resourceOrgId = (String) authorizationContext.getParameter("resourceOrgId");
+                if (StringUtils.isNotEmpty(resourceOrgId)) {
+                    String userResidentTenantDomain = user.getTenantDomain();
+                    OrganizationManager organizationManager = AuthorizationServiceHolder.getInstance().getOrganizationManager();
+                    String userResidentOrgId = organizationManager.resolveOrganizationId(userResidentTenantDomain);
+                    OrganizationUserSharingService organizationUserSharingService = new OrganizationUserSharingServiceImpl();
+                    List<UserAssociation> sharedAssociations = organizationUserSharingService.
+                            getUserAssociationsOfGivenUser(userId, userResidentOrgId);
+                    for (UserAssociation userAssociation : sharedAssociations) {
+                        if (resourceOrgId.equals(userAssociation.getOrganizationId())) {
+                            authenticatedUser.setAccessingOrganization(userAssociation.getOrganizationId());
+                            break;
+                        }
```

Comment:
Instead loading all shared user object from resident user, we can use 
`getUserAssociationOfAssociatedUserByOrgId(userId, resourceOrgId)`
Test and change